bindings/rust/s2n-tls/src/client_hello.rs

```diff
@@ -0,0 +1,401 @@
+use std::{
+    fmt,
+    ops::{Deref, DerefMut},
+};
+
+use s2n_tls_sys::*;
+
+use crate::error::{Error, Fallible};
+
+#[derive(Copy, Clone)]
+pub enum FingerprintType {
+    JA3,
+}
+
+// this is the size of the MD5 hash digest that is used for the JA3 fingerprint
+const MD5_HASH_SIZE: u32 = 16;
+
+impl From<FingerprintType> for s2n_tls_sys::s2n_fingerprint_type::Type {
+    fn from(value: FingerprintType) -> Self {
+        match value {
+            FingerprintType::JA3 => s2n_tls_sys::s2n_fingerprint_type::FINGERPRINT_JA3,
+        }
+    }
+}
+
+/// ClientHello is an opaque wrapper struct around `s2n_client_hello`. Note that
+/// the size of this type is not known, and as such it can only be used through
+/// references and pointers.
+///
+/// This implementation is motivated by the different memory_management required
+/// for different s2n_client_hello pointers. `s2n_client_hello_parse_message`
+/// returns a `*mut s2n_client_hello` which owns it's own data. This neatly fits
+///  the "smart pointer" pattern and can be represented as a `Box<T>`.
+///
+/// `s2n_connection_get_client_hello` returns a `*mut s2n_client_hello` which
+/// references memory owned by the connection, and therefore must not outlive
+/// the connection struct. This is best represented as a reference tied to the
+/// lifetime of the `Connection` struct.
+
+pub struct ClientHello(s2n_client_hello);
```

Comment:
This needs a `#[repr(transparent)]` if you're going to cast pointers

https://doc.rust-lang.org/1.26.2/unstable-book/language-features/repr-transparent.html

Reply:
I had actually looked at that, but I don't think it's allowed for this case. Since s2n_client_hello is an opaque struct, it has 0 size, and documentation call out that for `repr-tranparent`
> It also cannot be applied to enums, unions, empty structs, structs whose fields are all zero-sized, or structs with multiple non-zero-sized fields.

(from that link)